File: red/src/ssh.ts
// Application SSH arguments; colors-compute owns key lifecycle.
import { resolve } from "node:path";
import type { Opts } from "red/workflow";
import { keyMode } from "colors-compute-red";

export const buildPlaceholderDir = "/home/build-placeholder/.ssh";

export function renderedOnly(opts: Opts): boolean {
  return opts["red/event"] === "build" || Boolean(opts["red/dry-run"]);
}

// In keygen mode a build or dry-run names the placeholder identity rather
// than anything under the operator's home; a real event keeps whatever the
// library recorded.
export function withMachineKey(opts: Opts): Opts {
  if (keyMode(opts).mode !== "managed") return opts;
  const path = renderedOnly(opts) ? `${buildPlaceholderDir}/${opts.profile}` : opts["ssh-private-key-path"];
  if (path === undefined || path === null) return opts;
  return { ...opts, "ssh-private-key-path": path, "ssh-public-key-path": `${path}.pub` };
}

export function identityArgs(opts: Opts): string[] {
  const path = opts["ssh-private-key-path"];
  return path ? ["-i", String(path), "-o", "IdentitiesOnly=yes"] : [];
}

export function privateKeyPath(opts: Opts): string {
  if (!opts["ssh-private-key-path"]) throw new Error("deployment SSH identity unavailable");
  return resolve(String(opts["ssh-private-key-path"]));
}

File: blue/src/package_redis_blue/ssh.py
"""Application SSH arguments; colors-compute owns key lifecycle."""

from __future__ import annotations

import os

from colors_compute.ssh import _mode as ssh_mode

BUILD_PLACEHOLDER_DIR = "/home/build-placeholder/.ssh"


def rendered_only(opts: dict) -> bool:
    return opts.get("blue/event") == "build" or bool(opts.get("blue/dry-run"))


def with_machine_key(opts: dict) -> dict:
    if ssh_mode(opts)["mode"] != "managed":
        return opts
    path = f"{BUILD_PLACEHOLDER_DIR}/{opts.get('profile')}" if rendered_only(opts) else opts.get("ssh-private-key-path")
    if not path:
        return opts
    return {**opts, "ssh-private-key-path": path, "ssh-public-key-path": f"{path}.pub"}


def identity_args(opts: dict) -> list[str]:
    path = opts.get("ssh-private-key-path")
    return ["-i", path, "-o", "IdentitiesOnly=yes"] if path else []


def private_key_path(opts: dict) -> str:
    if not opts.get("ssh-private-key-path"):
        raise RuntimeError("deployment SSH identity unavailable")
    return os.path.abspath(str(opts["ssh-private-key-path"]))
